plugin/caddyfile/lexer.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2015 Light Code Labs, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddyfile
+
+import (
+	"bufio"
+	"io"
+	"unicode"
+)
+
+type (
+	// lexer is a utility which can get values, token by
+	// token, from a Reader. A token is a word, and tokens
+	// are separated by whitespace. A word can be enclosed
+	// in quotes if it contains whitespace.
+	lexer struct {
```

Comment:
@francislavoie @mholt
Had to copy lexer from caddy repo.
I'm fine with it, and lexer features don't change that often.

Does it make sense to expose in caddy a Parse function that preserves environment variables and snippets? Or maybe expose lexer publicly?

Reply:
Just to give more context:
- I parse several caddyfiles, from different sources.
- Then I merge them into a single caddyfile, putting global options first, then snippets then sites.

Because snippet declaration and import might be in different input caddyfiles, I must be able to do the parsing step without processing snippet and environment variables.

Reply:
I think this should do what you need? https://github.com/caddyserver/caddy/pull/3549

Reply:
Thanks. That's pretty much what I need.

I don't know the implications though, for example, by making it public probably caddy shouldn't make breaking changes on it. So, I'm completely fine if it's best for Caddy to keep the token concept an internal thing.

I will wait next caddy release before cleaning up duplicated code.

Reply:
Yeah for sure. I need Matt's +1 on it first.

Reply:
Dang, copying the whole lexer is going to be a pain to maintain. I've tried to avoid exporting it because then the burden is on us to make sure it doesn't change and then we have to maintain the API, and most people want the parser instead of the lexer anyway.

Your case is a bit niche... hmm. 

We might be able to get away with exporting it if we put a big scary warning in the godoc saying that this API is not stable.

Reply:
Indeed. Another option is to add a "parsePreservingSnippets" function. That parses it to server blocks, and each snippet would also be a "serverBlock".